File: frontend/src/types/inventory16.ts
/** Domain types for inventory / InventoryRecord16 */
export type InventoryRecord16Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface InventoryRecord16 {
  id: string;
  externalCode: string;
  displayName: string;
  status: InventoryRecord16Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface InventoryRecord16CreateRequest {
  externalCode: string;
  displayName: string;
  status: InventoryRecord16Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface InventoryRecord16ListResponse {
  items: InventoryRecord16[];
  total?: number;
}

export interface InventoryRecord16StatsResponse {
  activeCount: number;
}

export function isInventoryRecord16Active(rec: InventoryRecord16): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatInventoryRecord16Label(rec: InventoryRecord16): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareInventoryRecord16ByPriority(a: InventoryRecord16, b: InventoryRecord16): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/inventory_16.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// InventoryRecord16Service encapsulates business rules for InventoryRecord16.
type InventoryRecord16Service struct {
	repo *repository.InventoryRecord16Repository
}

// NewInventoryRecord16Service wires the service.
func NewInventoryRecord16Service(repo *repository.InventoryRecord16Repository) *InventoryRecord16Service {
	return &InventoryRecord16Service{repo: repo}
}

// Create validates and persists a new InventoryRecord16.
func (s *InventoryRecord16Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.InventoryRecord16, error) {
	now := time.Now().UTC()
	rec := &models.InventoryRecord16{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a InventoryRecord16 by id.
func (s *InventoryRecord16Service) Get(ctx context.Context, id string) (*models.InventoryRecord16, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped InventoryRecord16 pages.
func (s *InventoryRecord16Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.InventoryRecord16, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing InventoryRecord16.
func (s *InventoryRecord16Service) Update(ctx context.Context, actor string, rec *models.InventoryRecord16) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a InventoryRecord16.
func (s *InventoryRecord16Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds InventoryRecord16 by external code prefix.
func (s *InventoryRecord16Service) Search(ctx context.Context, facilityID, prefix string) ([]models.InventoryRecord16, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *InventoryRecord16Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
